### PR TITLE
De-conflate local from remote service names

### PR DESCRIPTION
**Pull request by codefromthecrypt**
Especially mentioned in #1794, local and remote service names are undeniably serving different use cases, yet currently conflated in the api.

/services -> only local serviceNames
/remoteServices?serviceName=X -> new: only remote serviceNames for auto-complete
/spans?remoteServiceName=X -> new: to restore functionality @llinder mentioned

cc @openzipkin/ui @bulicekj

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
oh wow.. cassandra tests pass now :) someone quick take a photo!

File: zipkin-storage/cassandra/src/main/java/zipkin2/storage/cassandra/Schema.java [hunk 1]
```diff
@@ -73,14 +75,22 @@ static Metadata readMetadata(Session session) {
     String compactionClass =
       keyspaceMetadata.getTable("span").getOptions().getCompaction().get("class");
 
-    return new Metadata(compactionClass);
+    return new Metadata(
+      compactionClass,
+      hasUpgrade1_autocompleteTags(keyspaceMetadata),
```
Comment (codefromthecrypt) [inline zipkin-storage/cassandra/src/main/java/zipkin2/storage/cassandra/Schema.java]: @zeagord you and I both forgot to check this!

File: zipkin-storage/cassandra/src/test/java/zipkin2/storage/cassandra/ITEnsureSchema.java [hunk 2]
```diff
@@ -77,12 +90,19 @@
     try (CassandraStorage storage = CassandraStorage.newBuilder()
       .contactPoints(contactPoint.getHostString() + ":" + contactPoint.getPort())
       .ensureSchema(false)
+      .autocompleteKeys(asList("environment"))
```
Comment (codefromthecrypt) [inline zipkin-storage/cassandra/src/test/java/zipkin2/storage/cassandra/ITEnsureSchema.java:98]: @zeagord here was the secret to causing us to figure out if we *really* work when schema isn't updated

**Comment by codefromthecrypt** [comment]
I am nearly done the hard parts. just have to implement query support for cassandra and cassandra v1. after that, it is easy

**Comment by codefromthecrypt** [comment]
@openzipkin/cassandra I'm a little stuck on the best way to model the remoteServiceName=X query.

I'm thinking about adding this for the remote service index. only other option I can think of is adding a SASI field on spans of r_service. thoughts?


```
CREATE TABLE IF NOT EXISTS zipkin2.trace_by_service_remote_service (
    service       text,             //-- service name, or blank to support queries without service name
    remote_service  text,             //-- remote service name
    bucket        int,              //-- time bucket, calculated as ts/interval (in microseconds), for some pre-configured interval like 1 day.
    ts            timeuuid,         //-- start timestamp of the span, truncated to millisecond precision
    trace_id      text,             //-- trace ID
    PRIMARY KEY ((service, remote_service, bucket), ts)
)
   WITH CLUSTERING ORDER BY (ts DESC)
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.TimeWindowCompactionStrategy'}
    AND default_time_to_live =  259200
    AND gc_grace_seconds = 3600
    AND read_repair_chance = 0
    AND dclocal_read_repair_chance = 0
    AND speculative_retry = '95percentile'
    AND comment = 'Secondary table for looking up a trace by remote service name. service column may be blank (when only looking up by remote_service). bucket column adds time bucketing to the partition key, values are microseconds rounded to a pre-configured interval (typically one day). ts column is start timestamp of the span as time-uuid, truncated to millisecond precision.';
```
Note: this would be in addition to [trace_by_service_span](https://github.com/openzipkin/zipkin/blob/master/zipkin-storage/cassandra/src/main/resources/zipkin2-schema-upgrade-1.cql), and require a client-side join if other parameters are sent (similar to how we do with annotation query).

Another option is to extend trace_by_service_span multiplying rows there by remote service name. This would be simpler to query... Not sure what's better? throwing this out to get feedback.

File: zipkin-storage/cassandra/src/test/java/zipkin2/storage/cassandra/ITEnsureSchema.java (hunk 2 above)
Comment (zeagord) [reply zipkin-storage/cassandra/src/test/java/zipkin2/storage/cassandra/ITEnsureSchema.java:98]: :) I totally forgot about this.

File: zipkin-storage/cassandra/src/main/resources/zipkin2-schema-upgrade-2.cql [hunk 3]
```diff
@@ -0,0 +1,30 @@
+CREATE TABLE IF NOT EXISTS zipkin2.remote_service_by_service (
```
Comment (codefromthecrypt) [inline zipkin-storage/cassandra/src/main/resources/zipkin2-schema-upgrade-2.cql:1]: @michaelsembwever @llinder @drolando can one of you have a look at this file? I'm just sort of replacing inline with what we did for span name indexing, but someone who knows cassandra should weigh in.

File: zipkin-storage/cassandra/src/main/resources/zipkin2-schema-indexes.cql [hunk 4]
```diff
@@ -33,20 +62,19 @@ CREATE TABLE IF NOT EXISTS zipkin2.span_by_service (
...
-    'analyzed': 'true',
-    'analyzer_class':'org.apache.cassandra.index.sasi.analyzer.DelimiterAnalyzer',
-    'delimiter': '░'};
-
-CREATE CUSTOM INDEX IF NOT EXISTS ON zipkin2.span (l_service) USING 'org.apache.cassandra.index.sasi.SASIIndex'
-   WITH OPTIONS = {'mode': 'PREFIX'};
-
-CREATE CUSTOM INDEX IF NOT EXISTS ON zipkin2.trace_by_service_span (duration) USING 'org.apache.cassandra.index.sasi.SASIIndex'
-   WITH OPTIONS = {'mode': 'PREFIX'};
+CREATE TABLE IF NOT EXISTS zipkin2.remote_service_by_service (
```
Comment (michaelsembwever) [inline zipkin-storage/cassandra/src/main/resources/zipkin2-schema-indexes.cql:65]: is adding the tables `trace_by_service_remote_service` and `remote_service_by_service` the only actual change in the file's diff?

File: zipkin-storage/cassandra/src/main/resources/zipkin2-schema-indexes.cql [hunk 5]
```diff
@@ -33,20 +62,19 @@ CREATE TABLE IF NOT EXISTS zipkin2.span_by_service (
     AND speculative_retry = '95percentile'
     AND comment = 'Secondary table for looking up span names by a service name.';
 
-DROP INDEX IF EXISTS zipkin2.span_annotation_query_idx;
```
Comment (codefromthecrypt) [inline zipkin-storage/cassandra/src/main/resources/zipkin2-schema-indexes.cql:36]: this index didn't exist

File: zipkin-storage/cassandra/src/main/resources/zipkin2-schema-upgrade-2.cql (hunk 3 above)
Comment (michaelsembwever) [reply zipkin-storage/cassandra/src/main/resources/zipkin2-schema-upgrade-2.cql:1]: It's an anti-pattern, you want to be careful with this, as it involves hot partitions. (That's why LCS with a low threshold and row caching. Deduping the writes is also important.)
So long as you apply the caution we took with `span_by_service` it's cool.

btw there's the stress profile if you want to see how the performance/throughput numbers differ…

File: zipkin-storage/cassandra/src/main/resources/zipkin2-schema-indexes.cql (hunk 4 above)
Comment (codefromthecrypt) [reply zipkin-storage/cassandra/src/main/resources/zipkin2-schema-indexes.cql:65]: yes, the other changes were moving the indexing near the tables they indexed as it was confusing me (at the cost of confusing you in this review.. sorry)

File: zipkin-storage/cassandra/src/main/resources/zipkin2-schema-upgrade-2.cql (hunk 3 above)
Comment (codefromthecrypt) [reply zipkin-storage/cassandra/src/main/resources/zipkin2-schema-upgrade-2.cql:1]: I will add comments to all the tables that have deduplication on them, as yeah in the schema now we make no such notes yet.

Comment (codefromthecrypt) [reply zipkin-storage/cassandra/src/main/resources/zipkin2-schema-upgrade-2.cql:1]: added a commit revising comments, feel free to suggest revisions to those if needed. thanks!

Comment (codefromthecrypt) [reply zipkin-storage/cassandra/src/main/resources/zipkin2-schema-upgrade-2.cql:1]: also backfilled missing stress tests

**Comment by codefromthecrypt** [comment]
cassandra 3.11.3+ done (except some unit tests which all around need to be added but saving for later)
next up cassandra 2.2+ impl (cassandra-v1)

**Comment by codefromthecrypt** [comment]
will merge this prior to the cassandra v1 impl as it was too hard to see what changed https://github.com/openzipkin/zipkin/pull/2492

**Comment by codefromthecrypt** [comment]
ok cassandra v1 "done" (same quality caveat)

now I will do a quality sweep including unit tests and README updates. if the build is green, it is testable

**Comment by codefromthecrypt** [comment]
only task left is unit test sweep.. doing that now.

**Pull request merged**